File: skills/persona-legal/scripts/internal/i18n/i18n.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package i18n provides internationalization and localization support.
package i18n

import (
	"encoding/json"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"

	"github.com/nicksnyder/go-i18n/v2/i18n"
	"golang.org/x/text/language"
)

var bundle *i18n.Bundle
var localizer *i18n.Localizer

// Init initializes the i18n bundle and localizer.
func Init(lang, defaultLang string, localesFS fs.FS) error {
	bundle = i18n.NewBundle(language.MustParse(defaultLang))
	bundle.RegisterUnmarshalFunc("json", json.Unmarshal)

	// Load all locale files from the FS
	files, err := fs.ReadDir(localesFS, ".")
	if err != nil {
		return fmt.Errorf("failed to read locales from FS: %w", err)
	}

	for _, f := range files {
		if filepath.Ext(f.Name()) == ".json" {
			data, err := fs.ReadFile(localesFS, f.Name())
			if err != nil {
				return fmt.Errorf("failed to read locale file %s: %w", f.Name(), err)
			}
			if _, err := bundle.ParseMessageFileBytes(data, f.Name()); err != nil {
				return fmt.Errorf("failed to parse locale file %s: %w", f.Name(), err)
			}
		}
	}

	if lang == "" {
		lang = os.Getenv("LANG")
	}

	localizer = i18n.NewLocalizer(bundle, lang, defaultLang)
	return nil
}

// T translates a simple string.
func T(id string) string {
	if localizer == nil {
		return id
	}
	msg, err := localizer.Localize(&i18n.LocalizeConfig{MessageID: id})
	if err != nil {
		return id
	}
	return msg
}

// TF translates a string with template variables.
func TF(id string, data map[string]interface{}) string {
	if localizer == nil {
		return id
	}
	msg, err := localizer.Localize(&i18n.LocalizeConfig{
		MessageID:    id,
		TemplateData: data,
	})
	if err != nil {
		return id
	}
	return msg
}

File: presentations/md2tty/src/i18n.js
/*!
 * @license
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/** @fileoverview i18n support */

const SUPPORTED_LANGS = ['en', 'es', 'fr', 'ru', 'zh', 'ar'];
let currentLangIndex = 0;

/** @type {Record<string, Record<string, string>>} */
const translations = {};

/**
 * Initializes i18n by fetching all supported languages upfront.
 */
export async function initI18n() {
  const urlParams = new URLSearchParams(window.location.search);
  const urlLang = urlParams.get('lang');
  const browserLang = (urlLang || navigator.language || 'en').substring(0, 2);
  currentLangIndex = SUPPORTED_LANGS.includes(browserLang) ? SUPPORTED_LANGS.indexOf(browserLang) : 0;

  const fetches = SUPPORTED_LANGS.map(lang => 
    fetch(`locales/${lang}.json`)
      .then(res => res.ok ? res.json() : [])
      .then(data => ({ lang, data }))
      .catch(e => {
        console.error(`Failed to load locale ${lang}`, e);
        return { lang, data: [] };
      })
  );

  const results = await Promise.all(fetches);
  
  for (const { lang, data } of results) {
    translations[lang] = {};
    for (const item of data) {
      translations[lang][item.id] = item.translation;
    }
  }
}

/**
 * Cycles to the next supported language.
 */
export function cycleLanguage() {
  currentLangIndex = (currentLangIndex + 1) % SUPPORTED_LANGS.length;
}

/**
 * Translates a key, replacing template variables like {{.Current}}
 * @param {string} key 
 * @param {Record<string, string|number>} params 
 * @returns {string}
 */
export function t(key, params = {}) {
  const currentLang = SUPPORTED_LANGS[currentLangIndex];
  const dict = translations[currentLang] || {};
  let str = dict[key] || (translations['en'] && translations['en'][key]) || key;
  for (const [k, v] of Object.entries(params)) {
    str = str.replace(`{{.${k}}}`, String(v));
  }
  return str;
}
